pkgs/applications/editors/neovim/wrapper.nix

```diff
@@ -55,44 +55,102 @@ let
 
   binPath = makeBinPath (optionals withRuby [rubyEnv] ++ optionals withNodeJs [nodejs]);
 
+  # Mapping a boolean argument to a key that tells us whether to add or not to
+  # add to nvim's 'embedded rc' this:
+  #
+  #    let g:<key>_host_prog=$out/bin/nvim-<key>
+  #
+  # Or this:
+  #
+  #    let g:loaded_${prog}_provider=1
+  #
+  # While the later tells nvim that this provider is not available
+  #
+  hostprog_check_table = {
+    node = withNodeJs;
+    python = withPython;
+    python3 = withPython3;
+    ruby = withRuby;
+  };
+  ## Here we calculate all of the arguments to the 1st call of `makeWrapper`
+  # We start with the executable itself NOTE we call this variable "initial"
+  # because if configure != {} we need to call makeWrapper twice, in order to
+  # avoid double wrapping, see comment near finalMakeWrapperArgs
+  initialMakeWrapperArgs = [ "${neovim}/bin/nvim" "${placeholder "out"}/bin/nvim" "--argv0" "$0" ]
+    ++ (lib.lists.flatten (lib.attrsets.mapAttrsToList (
+      # For every "host" program neovim supports, we map according to the attrset
+      # `hostprog_check_table` what flags to add to wrapper
+      prog:
+      withProg:
+      # --cmd is needed even if we disable this ${prog} provider or not
+      [ "--add-flags"
+        # This shell argument needs to be treated as 1 when calling makeWrapper
+        # that's why we use concatStringsSep
+        (lib.concatStringsSep " " [
+          "--cmd"
+          # Technically this escaping is for nvim
+          (lib.escapeShellArg (if withProg then
```

Comment:
This section can be simplified by only using a single `--add-flags` argument (uncommented):
```nix
{
  initialMakeWrapperArgs =
    let
      flags = lib.concatLists (lib.mapAttrsToList (prog: withProg: [
        "--cmd"
        (if withProg
        then "let g:${prog}_host_prog='${placeholder "out"}/bin/nvim-${prog}'"
        else "let g:loaded_${prog}_provider=1")
      ]) hostprog_check_table);
    in [
      "${neovim}/bin/nvim" "${placeholder "out"}/bin/nvim"
      "--argv0" "$0"
      "--add-flags" (lib.escapeShellArgs flags)
    ] ++ lib.optionals withRuby [
      "--set" "GEM_HOME" "${rubyEnv}/${rubyEnv.ruby.gemPath}"
    ] ++ lib.optionals (binPath != "") [
      "--suffix" "PATH" ":" binPath
    ];
}
```

Also note that generally there should never be a need to call `concatStringsSep " "` when you properly use `escapeShellArgs`, so I refactored it for that here.

Reply:
Thanks again for doing this. I took the liberty just to use this exactly but I changed the indentations a bit with my style.